src/batch-transaction.js

```diff
@@ -0,0 +1,343 @@
+ * @typedef {object} QueryPartition
+ * @property {string} partitionToken The partition token.
+ */
+/**
+ * @typedef {array} CreateQueryPartitionsResponse
+ * @property {QueryPartition[]} 0 List of query partitions.
+ * @property {object} 1 The full API response.
+ */
+/**
+ * @callback CreateQueryPartitionsCallback
+ * @param {?Error} err Request error, if any.
+ * @param {QueryPartition[]} partitions List of query partitions.
+ * @param {object} apiResponse The full API response.
+ */
+/**
+ * Creates a set of query partitions that can be used to execute a query
+ * operation in parallel. Partitions become invalid when the transaction used
+ * to create them is closed.
+ *
+ * @param {string|object} query A SQL query or
+ *     [`ExecuteSqlRequest`](https://cloud.google.com/spanner/docs/reference/rpc/google.spanner.v1#google.spanner.v1.ExecuteSqlRequest)
+ *     object.
+ * @param {object} [query.params] A map of parameter name to values.
+ * @param {object} [query.partitionOptions] A map of partition options.
+ * @param {object} [query.types] A map of parameter types.
+ * @param {CreateQueryPartitionsCallback} [callback] Callback function.
+ * @returns {Promise<CreateQueryPartitionsResponse>}
+ *
+ * @example <caption>include:samples/batch.js</caption>
+ * region_tag:create_query_partitions
+ */
+BatchTransaction.prototype.createQueryPartitions = function(query, callback) {
+  if (is.string(query)) {
+    query = {
+      sql: query,
+    };
+  }
+
+  var reqOpts = codec.encodeQuery(query);
+  var gaxOpts = query.gaxOptions;
+
+  if (gaxOpts) {
+    delete reqOpts.gaxOptions;
+  }
+
+  this.createPartitions_(
+    {
+      client: 'SpannerClient',
+      method: 'partitionQuery',
+      reqOpts,
+      gaxOpts,
+    },
+    callback
+  );
+};
+
+/**
+ * Generic create partition method. Handles common parameters used in both
+ * {@link BatchTransaction#createQueryPartitions} and {@link BatchTransaction#createReadPartitions}
+ *
+ * @private
+ *
+ * @param {object} config The request config.
+ * @param {function} Callback function.
+ */
+BatchTransaction.prototype.createPartitions_ = function(config, callback) {
+  var self = this;
+
+  var query = extend({}, config.reqOpts, {
+    session: this.session.formattedName_,
+    transaction: {id: this.id},
+  });
+
+  config.reqOpts = extend({}, query);
+  delete query.partitionOptions;
+
+  this.request(config, function(err, resp) {
+    if (err) {
+      callback(err, null, resp);
+      return;
+    }
+
+    var partitions = resp.partitions.map(function(partition) {
+      return extend({}, query, partition);
+    });
+
+    if (resp.transaction) {
+      self.id = resp.transaction.id;
+      self.readTimestamp = resp.transaction.readTimestamp;
+    }
+
+    callback(null, partitions, resp);
+  });
+};
+
+/**
+ * @typedef {object} ReadPartition
+ * @mixes ReadRequestOptions
+ * @property {string} partitionToken The partition token.
```

Comment:
This would normally be a place to do auto-unwrapping for the user, so they don't have to dig through an object with only one property (obviously, they want the token). As long as we add an `apiResponse` argument with the full, original object, it should be more convenient to the user:

```diff
var options = {
  table: 'Singers',
  keys: ['1'],
  columns: ['SingerId'],
}

- createReadPartitions(options, function(err, readPartitions) {
-   const readPartitionTokens = readPartitions.map(partition => partition.partitionToken)
- })
+ createReadPartitions(options, function(err, readPartitionTokens) {})
```

Reply:
I can't speak as to why its done this way, but in order to execute a partition you need the original query/read request as well. We actually deliberately extend the original query to include the partition token.

Reply:
> We actually deliberately extend the original query to include the partition token.

Where does that happen?

Reply:
Wait, sorry, pause this convo. I just realized this is the options argument and not the response from the API.

Reply:
*/mark as resolved*